From c2a34da24956f3d242ee2e11f25f7035b2bc07f4 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Sun, 6 Oct 2024 22:56:33 -0700
Subject: [PATCH 3/3] wifi: mac80211: Add percpu queues for utility functions

Use percpu SW queues and variables for all
the utility functions that involve stopping and waking up
the queues and add SKBs to this pending queue which is no
longer shared between different CPUs.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 net/mac80211/util.c    | 152 ++++++++++++++++++++++++++++++-----------
 1 file changed, 112 insertions(+), 40 deletions(-)

--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -398,18 +398,22 @@ void ieee80211_wake_txqs(struct tasklet_
 	int n_acs = IEEE80211_NUM_ACS;
 	unsigned long flags;
 	int i;
+	spinlock_t *queue_stop_reason_lock;
+	unsigned long *queue_stop_reasons;
 
 	rcu_read_lock();
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
+	queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	spin_lock_irqsave(queue_stop_reason_lock, flags);
 
 	if (local->hw.queues < IEEE80211_NUM_ACS)
 		n_acs = 1;
 
 	for (i = 0; i < local->hw.queues; i++) {
-		if (local->queue_stop_reasons[i])
+		queue_stop_reasons = this_cpu_ptr(local->queue_stop_reasons[i]);
+		if (*queue_stop_reasons)
 			continue;
 
-		spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+		spin_unlock_irqrestore(queue_stop_reason_lock, flags);
 		list_for_each_entry_rcu(sdata, &local->interfaces, list) {
 			int ac;
 
@@ -421,10 +425,10 @@ void ieee80211_wake_txqs(struct tasklet_
 					__ieee80211_wake_txqs(sdata, ac);
 			}
 		}
-		spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
+		spin_lock_irqsave(queue_stop_reason_lock, flags);
 	}
 
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	spin_unlock_irqrestore(queue_stop_reason_lock, flags);
 	rcu_read_unlock();
 }
 
@@ -433,32 +437,44 @@ static void __ieee80211_wake_queue(struc
 				   bool refcounted)
 {
 	struct ieee80211_local *local = hw_to_local(hw);
+	struct sk_buff_head *pcpu_pending;
+	unsigned long *queue_stop_reasons;
+	int *q_stop_reasons;
+	struct ieee80211_tasklet_data *pcpu_tx_pending_tasklet;
 
 	trace_wake_queue(local, queue, reason);
 
 	if (WARN_ON(queue >= hw->queues))
 		return;
 
-	if (!test_bit(reason, &local->queue_stop_reasons[queue]))
+	/* Get the current CPU's SW queue, tasklet struct and associated
+	 * variables specific to the given queue for the current CPU
+	 */
+	queue_stop_reasons = this_cpu_ptr(local->queue_stop_reasons[queue]);
+	q_stop_reasons = this_cpu_ptr(local->q_stop_reasons[queue][reason]);
+	pcpu_pending = this_cpu_ptr(local->pending[queue]);
+	pcpu_tx_pending_tasklet = this_cpu_ptr(local->tx_pending_tasklet);
+	if (!test_bit(reason, queue_stop_reasons))
 		return;
 
 	if (!refcounted) {
-		local->q_stop_reasons[queue][reason] = 0;
+		*q_stop_reasons = 0;
 	} else {
-		local->q_stop_reasons[queue][reason]--;
-		if (WARN_ON(local->q_stop_reasons[queue][reason] < 0))
-			local->q_stop_reasons[queue][reason] = 0;
+		(*q_stop_reasons)--;
+		if (WARN_ON(*q_stop_reasons < 0))
+			*q_stop_reasons = 0;
 	}
 
-	if (local->q_stop_reasons[queue][reason] == 0)
-		__clear_bit(reason, &local->queue_stop_reasons[queue]);
+	if (*q_stop_reasons == 0)
+		__clear_bit(reason, queue_stop_reasons);
 
-	if (local->queue_stop_reasons[queue] != 0)
+	if (*queue_stop_reasons != 0)
 		/* someone still has this queue stopped */
 		return;
 
-	if (!skb_queue_empty(&local->pending[queue]))
-		tasklet_schedule(&local->tx_pending_tasklet);
+	if (!skb_queue_empty(pcpu_pending)) {
+		tasklet_schedule(&pcpu_tx_pending_tasklet->tasklet);
+	}
 
 	/*
 	 * Calling _ieee80211_wake_txqs here can be a problem because it may
@@ -496,18 +512,22 @@ static void __ieee80211_stop_queue(struc
 				   bool refcounted)
 {
 	struct ieee80211_local *local = hw_to_local(hw);
+	unsigned long *queue_stop_reasons;
+	int *q_stop_reasons;
 
 	trace_stop_queue(local, queue, reason);
 
 	if (WARN_ON(queue >= hw->queues))
 		return;
 
+	queue_stop_reasons = this_cpu_ptr(local->queue_stop_reasons[queue]);
+	q_stop_reasons = this_cpu_ptr(local->q_stop_reasons[queue][reason]);
 	if (!refcounted)
-		local->q_stop_reasons[queue][reason] = 1;
+		*q_stop_reasons= 1;
 	else
-		local->q_stop_reasons[queue][reason]++;
+		(*q_stop_reasons)++;
 
-	set_bit(reason, &local->queue_stop_reasons[queue]);
+	set_bit(reason, queue_stop_reasons);
 }
 
 void ieee80211_stop_queue_by_reason(struct ieee80211_hw *hw, int queue,
@@ -537,19 +557,23 @@ void ieee80211_add_pending_skb(struct ie
 	unsigned long flags;
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	int queue = info->hw_queue;
+	spinlock_t *pcpu_queue_stop_reason_lock;
+	struct sk_buff_head *pcpu_pending;
 
 	if (WARN_ON(!info->control.vif)) {
 		ieee80211_free_txskb(&local->hw, skb);
 		return;
 	}
 
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
+	pcpu_queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	pcpu_pending = this_cpu_ptr(local->pending[queue]);
+	spin_lock_irqsave(pcpu_queue_stop_reason_lock, flags);
 	__ieee80211_stop_queue(hw, queue, IEEE80211_QUEUE_STOP_REASON_SKB_ADD,
 			       false);
-	__skb_queue_tail(&local->pending[queue], skb);
+	__skb_queue_tail(pcpu_pending, skb);
 	__ieee80211_wake_queue(hw, queue, IEEE80211_QUEUE_STOP_REASON_SKB_ADD,
 			       false);
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	spin_unlock_irqrestore(pcpu_queue_stop_reason_lock, flags);
 }
 
 void ieee80211_add_pending_skbs(struct ieee80211_local *local,
@@ -559,8 +583,11 @@ void ieee80211_add_pending_skbs(struct i
 	struct sk_buff *skb;
 	unsigned long flags;
 	int queue, i;
+	spinlock_t *pcpu_queue_stop_reason_lock;
+	struct sk_buff_head *pcpu_pending;
 
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
+	pcpu_queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	spin_lock_irqsave(pcpu_queue_stop_reason_lock, flags);
 	while ((skb = skb_dequeue(skbs))) {
 		struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 
@@ -570,19 +597,60 @@ void ieee80211_add_pending_skbs(struct i
 		}
 
 		queue = info->hw_queue;
+		pcpu_pending = this_cpu_ptr(local->pending[queue]);
 
 		__ieee80211_stop_queue(hw, queue,
 				IEEE80211_QUEUE_STOP_REASON_SKB_ADD,
 				false);
 
-		__skb_queue_tail(&local->pending[queue], skb);
+		__skb_queue_tail(pcpu_pending, skb);
 	}
 
 	for (i = 0; i < hw->queues; i++)
 		__ieee80211_wake_queue(hw, i,
 			IEEE80211_QUEUE_STOP_REASON_SKB_ADD,
 			false);
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	spin_unlock_irqrestore(pcpu_queue_stop_reason_lock, flags);
+}
+
+void ieee80211_unlink_all_skbs(struct ieee80211_sub_if_data *sdata)
+{
+	unsigned long flags;
+	struct ieee80211_local *local = sdata->local;
+	struct sk_buff *skb, *tmp;
+	spinlock_t *pcpu_queue_stop_reason_lock;
+	struct sk_buff_head *pcpu_pending;
+
+	pcpu_queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	spin_lock_irqsave(pcpu_queue_stop_reason_lock, flags);
+	for (int i = 0; i < IEEE80211_MAX_QUEUES; i++) {
+		pcpu_pending = this_cpu_ptr(local->pending[i]);
+		if (skb_queue_empty(pcpu_pending))
+			continue;
+		skb_queue_walk_safe(pcpu_pending, skb, tmp) {
+			struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
+			if (info->control.vif == &sdata->vif) {
+				__skb_unlink(skb, pcpu_pending);
+				ieee80211_free_txskb(&local->hw, skb);
+			}
+		}
+	}
+	spin_unlock_irqrestore(pcpu_queue_stop_reason_lock, flags);
+}
+
+void ieee80211_stop_queues_one_by_one(struct ieee80211_queue_info *queue_info)
+{
+	struct ieee80211_local *local = hw_to_local(queue_info->hw);
+	spinlock_t *pcpu_queue_stop_reason_lock;
+	unsigned long flags;
+
+	pcpu_queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	spin_lock_irqsave(pcpu_queue_stop_reason_lock, flags);
+
+	__ieee80211_stop_queue(queue_info->hw, queue_info->queue,
+			queue_info->reason, queue_info->refcounted);
+
+	spin_unlock_irqrestore(pcpu_queue_stop_reason_lock, flags);
 }
 
 void ieee80211_stop_queues_by_reason(struct ieee80211_hw *hw,
@@ -590,16 +658,22 @@ void ieee80211_stop_queues_by_reason(str
 				     enum queue_stop_reason reason,
 				     bool refcounted)
 {
-	struct ieee80211_local *local = hw_to_local(hw);
-	unsigned long flags;
 	int i;
+	struct ieee80211_queue_info queue_info;
 
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
-
-	for_each_set_bit(i, &queues, hw->queues)
-		__ieee80211_stop_queue(hw, i, reason, refcounted);
-
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	/* Stop all the queues of each cpu
+	 * We can do this by iterating over each AC and waking
+	 * up corresponding queue on all the CPUs
+	 *
+	 */
+	queue_info.hw = hw;
+	queue_info.reason = reason;
+	queue_info.refcounted = refcounted;
+
+	for_each_set_bit(i, &queues, hw->queues) {
+		queue_info.queue = i;
+		on_each_cpu((void (*)(void *))ieee80211_stop_queues_one_by_one, &queue_info, 1);
+	}
 }
 
 void ieee80211_stop_queues(struct ieee80211_hw *hw)
@@ -627,21 +701,42 @@ int ieee80211_queue_stopped(struct ieee8
 }
 EXPORT_SYMBOL(ieee80211_queue_stopped);
 
+void ieee80211_wake_queues_one_by_one(struct ieee80211_queue_info *queue_info)
+{
+	struct ieee80211_local *local = hw_to_local(queue_info->hw);
+	spinlock_t *pcpu_queue_stop_reason_lock;
+	unsigned long flags;
+
+	pcpu_queue_stop_reason_lock = this_cpu_ptr(local->queue_stop_reason_lock);
+	spin_lock_irqsave(pcpu_queue_stop_reason_lock, flags);
+
+	__ieee80211_wake_queue(queue_info->hw, queue_info->queue,
+			queue_info->reason, queue_info->refcounted);
+
+	spin_unlock_irqrestore(pcpu_queue_stop_reason_lock, flags);
+}
+
 void ieee80211_wake_queues_by_reason(struct ieee80211_hw *hw,
 				     unsigned long queues,
 				     enum queue_stop_reason reason,
 				     bool refcounted)
 {
-	struct ieee80211_local *local = hw_to_local(hw);
-	unsigned long flags;
 	int i;
+	struct ieee80211_queue_info queue_info;
 
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
-
-	for_each_set_bit(i, &queues, hw->queues)
-		__ieee80211_wake_queue(hw, i, reason, refcounted);
-
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	/* Wake up all the 4 queues of each cpu
+	 * We can do this by iterating over each AC and waking
+	 * up corresponding queue on all the CPUs
+	 *
+	 */
+	queue_info.hw = hw;
+	queue_info.reason = reason;
+	queue_info.refcounted = refcounted;
+
+	for_each_set_bit(i, &queues, hw->queues) {
+		queue_info.queue = i;
+		on_each_cpu((void (*)(void *))ieee80211_wake_queues_one_by_one, &queue_info, 1);
+	}
 }
 
 void ieee80211_wake_queues(struct ieee80211_hw *hw)
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -2489,6 +2489,7 @@ void ieee80211_flush_queues(struct ieee8
 void __ieee80211_flush_queues(struct ieee80211_local *local,
 			      struct ieee80211_sub_if_data *sdata,
 			      unsigned int queues, bool drop);
+void ieee80211_unlink_all_skbs(struct ieee80211_sub_if_data *sdata);
 
 static inline bool ieee80211_can_run_worker(struct ieee80211_local *local)
 {
--- a/net/mac80211/iface.c
+++ b/net/mac80211/iface.c
@@ -659,17 +659,10 @@ static void ieee80211_do_stop(struct iee
 		skb_queue_purge(&sdata->status_queue);
 	}
 
-	spin_lock_irqsave(&local->queue_stop_reason_lock, flags);
-	for (i = 0; i < IEEE80211_MAX_QUEUES; i++) {
-		skb_queue_walk_safe(&local->pending[i], skb, tmp) {
-			struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
-			if (info->control.vif == &sdata->vif) {
-				__skb_unlink(skb, &local->pending[i]);
-				ieee80211_free_txskb(&local->hw, skb);
-			}
-		}
-	}
-	spin_unlock_irqrestore(&local->queue_stop_reason_lock, flags);
+	/* Since there are percpu SW queues, unlink all
+	 * the skbs on each CPU SW queues
+	 */
+	on_each_cpu((void (*)(void *))ieee80211_unlink_all_skbs, sdata, 1);
 
 	if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN)
 		ieee80211_txq_remove_vlan(local, sdata);
@@ -677,7 +670,7 @@ static void ieee80211_do_stop(struct iee
 	sdata->bss = NULL;
 
 	if (local->open_count == 0)
-		ieee80211_clear_tx_pending(local);
+		on_each_cpu((void (*)(void *))ieee80211_clear_tx_pending, local, 1);
 
 	sdata->vif.bss_conf.beacon_int = 0;
 
--- a/net/mac80211/main.c
+++ b/net/mac80211/main.c
@@ -1653,8 +1653,13 @@ EXPORT_SYMBOL(ieee80211_register_hw);
 void ieee80211_unregister_hw(struct ieee80211_hw *hw)
 {
 	struct ieee80211_local *local = hw_to_local(hw);
+	int cpu;
 
-	tasklet_kill(&local->tx_pending_tasklet);
+	for_each_possible_cpu(cpu) {
+		struct ieee80211_tasklet_data *tasklet_data =
+			 per_cpu_ptr(local->tx_pending_tasklet, cpu);
+		tasklet_kill(&tasklet_data->tasklet);
+	}
 	tasklet_kill(&local->tasklet);
 
 #ifdef __disabled__CONFIG_INET
@@ -1685,7 +1690,7 @@ void ieee80211_unregister_hw(struct ieee
 	cancel_work_sync(&local->restart_work);
 	cancel_work_sync(&local->reconfig_filter);
 
-	ieee80211_clear_tx_pending(local);
+	on_each_cpu((void (*)(void *))ieee80211_clear_tx_pending, local, 1);
 	rate_control_deinitialize(local);
 
 	if (skb_queue_len(&local->skb_queue) ||
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -5138,9 +5138,11 @@ void ieee80211_clear_tx_pending(struct i
 {
 	struct sk_buff *skb;
 	int i;
+	struct sk_buff_head *pcpu_pending;
 
 	for (i = 0; i < local->hw.queues; i++) {
-		while ((skb = skb_dequeue(&local->pending[i])) != NULL)
+		pcpu_pending = this_cpu_ptr(local->pending[i]);
+		while ((skb = skb_dequeue(pcpu_pending)) != NULL)
 			ieee80211_free_txskb(&local->hw, skb);
 	}
 }
--- a/net/mac80211/tdls.c
+++ b/net/mac80211/tdls.c
@@ -1445,7 +1445,7 @@ int ieee80211_tdls_oper(struct wiphy *wi
 	struct sta_info *sta;
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
 	struct ieee80211_local *local = sdata->local;
-	int ret;
+	int ret, cpu;
 
 	if (!(wiphy->flags & WIPHY_FLAG_SUPPORTS_TDLS))
 		return -ENOTSUPP;
@@ -1508,7 +1508,11 @@ int ieee80211_tdls_oper(struct wiphy *wi
 		 * Note that this only forces the tasklet to flush pendings -
 		 * not to stop the tasklet from rescheduling itself.
 		 */
-		tasklet_kill(&local->tx_pending_tasklet);
+		for_each_possible_cpu(cpu) {
+			struct ieee80211_tasklet_data *tasklet_data =
+				 per_cpu_ptr(local->tx_pending_tasklet, cpu);
+			tasklet_kill(&tasklet_data->tasklet);
+		}
 		/* flush a potentially queued teardown packet */
 		ieee80211_flush_queues(local, sdata, false);
 
